src/core/coerce.c

```diff
@@ -332,11 +333,131 @@ MVMString * MVM_coerce_n_s(MVMThreadContext *tc, MVMnum64 n) {
     return MVM_string_ascii_from_buf_nocheck(tc, blob, orig_len);
 }
 
-MVMint64 MVM_coerce_s_i(MVMThreadContext *tc, MVMString *s) {
-    char     *enc = MVM_string_ascii_encode_any(tc, s);
-    MVMint64  i   = strtoll(enc, NULL, 10);
-    MVM_free(enc);
-    return i;
+MVMint64 MVM_coerce_s_i(MVMThreadContext *tc, MVMString *str) {
+    MVMStringIndex strgraphs = MVM_string_graphs(tc, str);
+    MVMint64       result = 0;
+    MVMint32       any = 0, negative = 0;
+
+    signed long long cutoff;
+    MVMint32  cutlim;
+
+    if (!strgraphs)
+        return result;
+
+/*-
+ * Copyright (c) 1990 The Regents of the University of California.
+ * with minor modifications to simplify and work in MoarVM
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ * 3. [rescinded 22 July 1999]
+ * 4. Neither the name of the University nor the names of its contributors
+ *    may be used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
+ * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ */
+    if (str->body.storage_type == MVM_STRING_GRAPHEME_ASCII) {
+        const MVMGraphemeASCII *s = str->body.storage.blob_ascii;
+        MVMStringIndex i = 0;
+        MVMGraphemeASCII c;
+        do {
+            c = *s++;
+        } while (i++ < strgraphs && isspace(c));
```

Comment:
I don't think we want to use `isspace` (or `isdigit` below). Or anything from `<ctype.h>` as all the functions are sensitive to C locales. IIRC the obscure trap here is that even if (well, "as") we don't call  `setlocale` anywhere, we're (still) subject to the "C" locale, which is only defined for ASCII. *Most* everywhere ends up treating the octets 128-255 as false for all `is*` functions, but (at least) VMS manages to have some return true, because (I inferred, remotely) that it was treating those octets as ISO-8859-1 (or whichever DEC character set forms the basis for ISO-8859-1). So presumably other platforms could do this (and have `isspace(0xA0)` return true, or any other implementation defined behaviour they care to torture us with)

So, I think we need to implement our own here, which really are hard coded for ASCII. In turn, I suspect that for speed, `isgraph` likely needs to be a lookup table, but `isdigit` I guess is comparisons `... >= '0' && ... <= '9'`.

Reply:
I don't have strong feelings either way, but since we're currently just using `strtoll()` from stdlib.h, aren't we already locale sensitive?

Reply:
Yes, that's a good point. We are.

I might view that as a bug :-)

Which would mean that it isn't troubling me that this is a (theoretical) "regression" and we leave the code as-is. For now, and explore whether replacing them in a future commit with something hard-coded to ASCII gives slightly better compiled code.